src/hashset.c

```diff
@@ -0,0 +1,1705 @@
+                iter->index = iter->hashset->rehashIdx;
+            }
+            iter->posInBucket = 0;
+        } else {
+            /* Advance position within bucket, or bucket index, or table. */
+            iter->posInBucket++;
+            if (iter->posInBucket >= ELEMENTS_PER_BUCKET) {
+                iter->posInBucket = 0;
+                iter->index++;
+                if (iter->index >= (long)numBuckets(iter->hashset->bucketExp[iter->table])) {
+                    iter->index = 0;
+                    if (hashsetIsRehashing(iter->hashset) && iter->table == 0) {
+                        iter->table++;
+                    } else {
+                        /* Done. */
+                        break;
+                    }
+                }
+            }
+        }
+        bucket *b = &iter->hashset->tables[iter->table][iter->index];
+        if (!(b->presence & (1 << iter->posInBucket))) {
+            /* No element here. Skip. */
+            continue;
+        }
+        /* Return the element at this position. */
+        if (elemptr) {
+            *elemptr = b->elements[iter->posInBucket];
+        }
+        return 1;
+    }
+    return 0;
+}
+
+/* --- Random elements --- */
+
+/* Points 'found' to a random element in the hash table and returns 1. Returns 0
+ * if the table is empty. */
+int hashsetRandomElement(hashset *t, void **found) {
+    void *samples[WEAK_RANDOM_SAMPLE_SIZE];
+    unsigned count = hashsetSampleElements(t, (void **)&samples, WEAK_RANDOM_SAMPLE_SIZE);
+    if (count == 0) return 0;
+    unsigned idx = random() % count;
+    *found = samples[idx];
+    return 1;
+}
+
+/* Points 'found' to a random element in the hash table and returns 1. Returns 0
+ * if the table is empty. This one is more fair than hashsetRandomElement(). */
+int hashsetFairRandomElement(hashset *t, void **found) {
+    void *samples[FAIR_RANDOM_SAMPLE_SIZE];
+    unsigned count = hashsetSampleElements(t, (void **)&samples, FAIR_RANDOM_SAMPLE_SIZE);
+    if (count == 0) return 0;
+    unsigned idx = random() % count;
+    *found = samples[idx];
+    return 1;
+}
+
+/* This function samples a sequence of elements starting at a random location in
+ * the hash table.
+ *
+ * The sampled elements are stored in the array 'dst' which must have space for
+ * at least 'count' elements.te
+ *
+ * The function returns the number of sampled elements, which is 'count' except
+ * if 'count' is greater than the total number of elements in the hash table. */
+unsigned hashsetSampleElements(hashset *t, void **dst, unsigned count) {
+    /* Adjust count. */
+    if (count > hashsetSize(t)) count = hashsetSize(t);
+    scan_samples samples;
+    samples.size = count;
+    samples.count = 0;
+    samples.elements = dst;
+    size_t cursor = randomSizeT();
+    while (samples.count < count) {
+        cursor = hashsetScan(t, cursor, sampleElementsScanFn, &samples, HASHSET_SCAN_SINGLE_STEP);
+    }
+    rehashStepOnReadIfNeeded(t);
+    return count;
+}
```

Comment:
Optimization: Select the sample index before sampling instead of after. Then just count up elements without copying in `sampleElementsScanFn()` and end the loop early when the random index is reached, and return that element. Space complexity goes from O(count) -> O(1) and can be stack allocated, and average time complexity is 2x faster.

I can write a PR for this. :) The logic is equivalent but maybe the fairness would be less intuitive.

Reply:
Good idea. PR targeting the hashset branch please.

Reply:
But I believe sample elements is used separately for eviction, via kvstoreHashsetSampleElements, so we still need the sample elements function too.

It's a balance between complexity and efficiency. For hashset random, the array is stack allocated and copying just the pointers isn't much overhead since the stack memory is already in CPU cache. So I'm not sure if it's worth it or not. See also Madelyn's comment on the hashset PR for hashset non-fair random.